### PR TITLE
Upgrade styler to better handle lists

### LINKED ISSUE #9408: [autoformatters] wrapping destroying items/lists
# 🚀 Feature request

Would it be possible to make the auto-wrappers respect items/lists?

e.g. ended up with:
```
        """
        number of training steps is either 1. args.max_steps if --max_steps > 1 2. else derive from dataset if we can
        get its size
        """
```
not only it's broken, it's unreadable.

The original was:
```
        """
        number of training steps is either 
        1. args.max_steps if --max_steps > 1 
        2. else derive from dataset if we can get its size
        """
```

Ideally it should not remove new lines before bullets */- and numbers 1. 

I also am not sure why there is a need to merge lines when the writer meant them to be shorter. I get the shortening, but why can't short lines be left alone. Which would be the case in this example. It looks like the only way I can enforce readable content is to inject paragraphs.

Thank you!

@sgugger

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR upgrades the doc styling script to automatically add new lines before lists. This makes the script more robust as it will avoid reformatting those lists and make them appear properly once sphinx has done its thing.

In passing a few badly formatted docstrings/doc pages are fixed, just waiting for some input from @patrickvonplaten for the problems in LED/Longformer.

Fixes #9408

### CODE REVIEW

File: src/transformers/models/layoutlm/modeling_layoutlm.py [hunk 1]
```diff
@@ -638,8 +638,8 @@ def _init_weights(self, module):
 
             `What are input IDs? <../glossary.html#input-ids>`__
         bbox (:obj:`torch.LongTensor` of shape :obj:`({0}, 4)`, `optional`):
-            Bounding Boxes of each input sequence tokens. Selected in the range ``[0, config.max_2d_position_embeddings
-            - 1]``.
```
Comment (sgugger) [inline src/transformers/models/layoutlm/modeling_layoutlm.py:642]: This is the only real failure of the upgraded script on the whole lib.
With the upgrade, the script now thinks that line `   - 1` is the start of a new list. I think this is an acceptable price to pay as there is an easy fix for those cases (just remove the space between the - and the previous word).

File: src/transformers/models/led/modeling_led.py [hunk 2]
```diff
@@ -174,8 +174,7 @@ def forward(
 
         The `attention_mask` is changed in :meth:`BertModel.forward` from 0, 1, 2 to -ve: no attention
 
-              0: local attention
-            +ve: global attention
+              0: local attention +ve: global attention
```
Comment (sgugger) [inline src/transformers/models/led/modeling_led.py:177]: Not sure how to format those as I don't understand this docstring.

File: src/transformers/models/layoutlm/modeling_layoutlm.py (hunk 1 above)
Comment (stas00) [reply src/transformers/models/layoutlm/modeling_layoutlm.py:642]: could it lookeahed and see if there are more items? If there none, then it is not a list.

File: src/transformers/models/led/modeling_led.py (hunk 2 above)
Comment (stas00) [reply src/transformers/models/led/modeling_led.py:177]: add `- ` bullet points?
```
-  0: local attention
- +ve: global attention

File: src/transformers/models/layoutlm/modeling_layoutlm.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/models/layoutlm/modeling_layoutlm.py:642]: The problem is that this logic is a bit complex to implement since the next item may be several lines ahead (since one item can take several lines). It's definitely possible, but I would like to only add it if we really need it since it will add some extra complexity (and thus execution time).

File: src/transformers/models/led/modeling_led.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/models/led/modeling_led.py:177]: I can do that, but +ve does not make any sense to me, so I don't know what was the original author's idea behind that docstring. Asked patrick since he wrote that code.

File: src/transformers/models/layoutlm/modeling_layoutlm.py (hunk 1 above)
Comment (stas00) [reply src/transformers/models/layoutlm/modeling_layoutlm.py:642]: It probably depends on how it's implemented. 

If it's a single paragraph we are looking at then it can be parsed directly into an optional header and a bullet list, then you immediately know whether there are multiple bullets or just one. so a hypothetical parser could go:
```
header, items = parse_to_bullet_list(para)
if len(items) < 2: this is not a bullet list.
```

Of course, this could still fail if the input happened to be:
```
Bounding Boxes of each input sequence tokens. Selected in the range ``[0, config.max_2d_position_embeddings
            - 1]``. blah blahof each input sequence tokens. Selected in the range ``[0, something_else
            - 5]``. 
```
So there is no winning it seems.

**Review by stas00 — APPROVED** [review]
Looks fantastic, @sgugger - I did some experiments and it even indents to the item the long list entries when it wraps those long line. Very awesome!

**Review by LysandreJik — APPROVED** [review]
Very cool!!

**Comment by stas00** [comment]
I suppose I need to stick to the common bullet format, as it couldn't handle this `\d\) ` style of bullets.
```
-1) Optimizer State Partitioning (stage 1)
-2) Add Gradient Partitioning (stage 2)
+1) Optimizer State Partitioning (stage 1) 2) Add Gradient Partitioning (stage 2)
```

**Comment by stas00** [comment]
OK, not sure about this one. In .rst I had:

```
Miscellaneous notes:
- DeepSpeed works with the PyTorch Trainer but not TF Trainer.
- While DeepSpeed has a pip installable PyPI package, 
```
the style wrapper broke the bullets and made them into one paragraph/line.
```
Miscellaneous notes: - DeepSpeed works with the PyTorch Trainer but not TF Trainer. - While DeepSpeed has a pip installable PyPI package, 
```
Is `-` an illegal bullet in sphinx? If so could it be automatically swapped for a legal one? `-` works in markdown/github/etc.

Thank.

**Comment by sgugger** [comment]
Mmm, the patch should be applied to the rst files too (can't link to the diff but it's line 384 of the last file in the diff shown by GitHub).

**Comment by stas00** [comment]
I re-based just in case, and no, it still doesn't insert the line. Here is the exact para:
```
Miscellaneous notes:
* DeepSpeed works with the PyTorch Trainer but not TF Trainer.
* While DeepSpeed has a pip installable PyPI package, it is highly recommended that it be `installed from source
  <https://github.com/microsoft/deepspeed#installation>`__ to best match your hardware and also to enable features like
  1-bit Adam, which aren't available in the pypi distribution.
```

**Comment by sgugger** [comment]
Indeed, I made some stupid mistake, #9488 should fix this.
